pkg/controller/backupbucket/actuator.go

```diff
@@ -6,43 +6,162 @@ package backupbucket
+	awsclient "github.com/gardener/gardener-extension-provider-aws/pkg/aws/client"
 )
 
 type actuator struct {
 	backupbucket.Actuator
-	client client.Client
+	client           client.Client
+	awsClientFactory awsclient.Factory
 }
 
-func newActuator(mgr manager.Manager) backupbucket.Actuator {
+// NewActuator creates a new Actuator that creates/updates backup-bucket.
+func NewActuator(mgr manager.Manager, awsClientFactory awsclient.Factory) backupbucket.Actuator {
 	return &actuator{
-		client: mgr.GetClient(),
+		client:           mgr.GetClient(),
+		awsClientFactory: awsClientFactory,
 	}
 }
 
-func (a *actuator) Reconcile(ctx context.Context, _ logr.Logger, bb *extensionsv1alpha1.BackupBucket) error {
-	awsClient, err := aws.NewClientFromSecretRef(ctx, a.client, bb.Spec.SecretRef, bb.Spec.Region)
+// Reconcile reconciles the BackupBucket resource with following steps:
+// 1. Create aws client from secret ref.
+// 2. Decode the backupbucket config (if provided).
+// 3. Check if bucket already exist or not.
+// 4. If bucket doesn't exist
+//   - then create a new bucket according to backupbucketConfig, if provided.
+//
+// 5. If bucket exist
+//   - check for bucket update is required or not
+//   - If yes then update the backup bucket settings according to backupbucketConfig(if provided)
+//     otherwise do nothing.
+func (a *actuator) Reconcile(ctx context.Context, logger logr.Logger, bb *extensionsv1alpha1.BackupBucket) error {
+	logger.Info("Starting reconciliation of BackupBucket...")
+
+	authConfig, err := aws.GetCredentialsFromSecretRef(ctx, a.client, bb.Spec.SecretRef, false, bb.Spec.Region)
+	if err != nil {
+		return util.DetermineError(fmt.Errorf("could not get AWS credentials: %w", err), helper.KnownCodes)
+	}
+
+	awsClient, err := a.awsClientFactory.NewClient(*authConfig)
 	if err != nil {
 		return util.DetermineError(err, helper.KnownCodes)
 	}
 
-	return util.DetermineError(awsClient.CreateBucketIfNotExists(ctx, bb.Name, bb.Spec.Region), helper.KnownCodes)
+	var backupbucketConfig *apisaws.BackupBucketConfig
+	if bb.Spec.ProviderConfig != nil {
+		backupbucketConfig, err = validator.DecodeBackupBucketConfig(serializer.NewCodecFactory(a.client.Scheme(), serializer.EnableStrict).UniversalDecoder(), bb.Spec.ProviderConfig)
+		if err != nil {
+			logger.Error(err, "Failed to decode provider config")
+			return err
+		}
+	}
+
+	bucketVersioningStatus, err := awsClient.GetBucketVersioningStatus(ctx, bb.Name)
+	if err != nil {
+		var apiErr smithy.APIError
+		if errors.As(err, &apiErr) {
+			switch apiErr.ErrorCode() {
+			case "NoSuchBucket":
+				// bucket doesn't exist, create the bucket with buckupbucket config (if provided)
+				return util.DetermineError(awsClient.CreateBucket(ctx, bb.Name, bb.Spec.Region, backupbucketConfig), helper.KnownCodes)
+			case "PermanentRedirect":
+				return util.DetermineError(fmt.Errorf("bucket exists in different region"), helper.KnownCodes)
```

Comment:
Maybe you can return exactly the error that you want here (if you expect to return a particular error code) instead of going sideways by producing a string that may result in the error you want.